examples/generative/cyclegan.py

```diff
@@ -0,0 +1,695 @@
+
+            # Discriminator loss
+            disc_X_loss = self.discriminator_loss_fn(disc_real_x, disc_fake_x)
+            disc_Y_loss = self.discriminator_loss_fn(disc_real_y, disc_fake_y)
+
+        # Get the gradients for the generators
+        grads_G = tape.gradient(total_loss_G, self.gen_G.trainable_variables)
+        grads_F = tape.gradient(total_loss_F, self.gen_F.trainable_variables)
+
+        # Get the gradients for the discriminators
+        disc_X_grads = tape.gradient(disc_X_loss, self.disc_X.trainable_variables)
+        disc_Y_grads = tape.gradient(disc_Y_loss, self.disc_Y.trainable_variables)
+
+        # Update the weights of the generators
+        self.gen_G_optimizer.apply_gradients(
+            zip(grads_G, self.gen_G.trainable_variables)
+        )
+        self.gen_F_optimizer.apply_gradients(
+            zip(grads_F, self.gen_F.trainable_variables)
+        )
+
+        # Update the weights of the discriminators
+        self.disc_X_optimizer.apply_gradients(
+            zip(disc_X_grads, self.disc_X.trainable_variables)
+        )
+        self.disc_Y_optimizer.apply_gradients(
+            zip(disc_Y_grads, self.disc_Y.trainable_variables)
+        )
+
+        return {
+            "G_loss": total_loss_G,
+            "F_loss": total_loss_F,
+            "D_X_loss": disc_X_loss,
+            "D_Y_loss": disc_Y_loss,
+        }
+
+
+"""
+## Create a callback that periodically saves generated images
+"""
+
+
+class GANMonitor(keras.callbacks.Callback):
+    """A callback to generate and save images after each epoch"""
+
+    def __init__(self, num_img=4):
+        self.num_img = num_img
+
+    def on_epoch_end(self, epoch, logs=None):
+        _, ax = plt.subplots(4, 2, figsize=(12, 12))
+        for i, img in enumerate(test_horses.take(4)):
+            prediction = self.model.gen_G(img)[0].numpy()
+            prediction = (prediction * 127.5 + 127.5).astype(np.uint8)
+            img = (img[0] * 127.5 + 127.5).numpy().astype(np.uint8)
+
+            ax[i, 0].imshow(img)
+            ax[i, 1].imshow(prediction)
+            ax[i, 0].set_title("Input image")
+            ax[i, 1].set_title("Translated image")
+            ax[i, 0].axis("off")
+            ax[i, 1].axis("off")
+
+            prediction = keras.preprocessing.image.array_to_img(prediction)
+            prediction.save(
+                "generated_img_{i}_{epoch}.png".format(i=i, epoch=epoch + 1)
+            )
+        plt.show()
+        plt.close()
+
+
+"""
+## Create a callback that periodically saves the models and optimizers
+"""
+
+
+class ModelCheckpointer(keras.callbacks.Callback):
+    """Saves model weights and optimizers"""
+
+    def __init__(self, checkpoint_path, save_freq=5):
+        self.checkpoint_path = checkpoint_path
+        self.save_freq = save_freq
+
+    def on_epoch_end(self, epoch, logs=None):
+        if epoch == 0:
+            ckpt = tf.train.Checkpoint(
```

Comment:
Why do you need this setup -- wouldn't `model.save()` just work?

Reply:
`model.save()` doesn't work in this case.

Reply:
What's the problem?

Reply:
Give me some time. I will double check this. I think I made a mistake while checking for this option.